### PR TITLE
[Feature] Bookmark pool candidates

### LINKED ISSUE #8598: ✨ Bookmark a pool candidate
## ✨ Feature

Pool operators would like a way to bookmark a pool candidate that they want to keep an eye on.

## 🎨 Design File

by @Jerryescandon 
https://www.figma.com/file/em5ArPr044YaIe1Vb1zCZz/Pools---Record-of-Decision?type=design&node-id=6872-37861&mode=design&t=7j6ZyDxN1M9FKKor-0

## 📸 Screenshot

image

## 🙋‍♀️ Proposed Implementation

Add a boolean isBookmarked field to the PoolCandidate query types and mutations.  Read and write should be restricted in the same way as other "manager fields" like notes.

Suggestion/note: Mutation that returns boolean to avoid reloads. Caution advised 

## ✅ Acceptance Criteria

- [ ] field added to database and API
- [ ] Proper restrictions on access and mutation
- [ ] Shows up in the assessment tracker
- [ ] Sorts row to top of assessment tracker when bookmarked
  -  isBookmarked
  -  Decision
  -  isPriority
  -  isVeterant
  -  alpha by last name
  -  alpha by first name
- [ ] Clicking icon in assessment tracker toggles bookmarked status
  - [ ] Prevent the table from reloading and causing a re-sort

### PR DESCRIPTION
**Pull request by JamesHuf**
🤖 Resolves #8598

## 👋 Introduction

Adds the ability to bookmark pool candidates in the assessment step tracker.

## 🧪 Testing

The assessment step tracker is currently not used on any page so testing must be done in storybook.

1. Manually test the new mutation and confirm it works as expected
```
mutation ToggleBookmark_Mutation($id: ID!) {
    togglePoolCandidateBookmark(id: $id)
}
```
2. Open the assessment step tracker story
3. Confirm already bookmarked candidates are displayed at the top of their columns.
4. Click to bookmark a candidate and confirm they are NOT immediately re-sorted to the top of the column.

## 📸 Screenshot

image

### CODE REVIEW

File: apps/web/src/components/AssessmentStepTracker/AssessmentResults.tsx [hunk 1]
```diff
@@ -80,6 +95,48 @@ const AssessmentResult = ({
...
+                    "Alert displayed to the user when they un-mark a a candidate as bookmarked.",
+                }),
+              );
+            }
+          }
+        })
+        .catch(() => {
+          toast.error(
+            intl.formatMessage({
+              defaultMessage: "Error: failed updating a candidates bookmark.",
```
Comment (mnigh) [inline apps/web/src/components/AssessmentStepTracker/AssessmentResults.tsx:130]:
```suggestion
              defaultMessage: "Error: failed updating a candidate's bookmark.",
```

Comment (JamesHuf) [reply apps/web/src/components/AssessmentStepTracker/AssessmentResults.tsx:130]: Apostrophes are over-rated!

File: apps/web/src/components/AssessmentStepTracker/AssessmentResults.tsx [hunk 2]
```diff
@@ -88,19 +145,72 @@ const AssessmentResult = ({
...
+          />
+        ) : (
+          <Button
+            mode="icon_only"
+            color="black"
+            onClick={toggleBookmark}
+            icon={BookmarkIconOutline}
+            aria-label={intl.formatMessage(
+              {
+                defaultMessage: "Bookmark {candidateName} to top of column.",
```
Comment (mnigh) [inline apps/web/src/components/AssessmentStepTracker/AssessmentResults.tsx:188]: Isn't the verb here meant to be **pin**?

File: apps/web/src/components/AssessmentStepTracker/AssessmentResults.tsx [hunk 3]
```diff
@@ -88,19 +145,72 @@ const AssessmentResult = ({
...
+        {isBookmarked ? (
+          <Button
+            mode="icon_only"
+            color="primary"
+            onClick={toggleBookmark}
+            icon={BookmarkIconSolid}
+            aria-label={intl.formatMessage(
+              {
+                defaultMessage:
+                  "Un-bookmark {candidateName} from top of column.",
```
Comment (mnigh) [inline apps/web/src/components/AssessmentStepTracker/AssessmentResults.tsx:162]: Isn't the verb here meant to be **pin**?

File: apps/web/src/components/AssessmentStepTracker/AssessmentResults.tsx (hunk 2 above)
Comment (JamesHuf) [reply apps/web/src/components/AssessmentStepTracker/AssessmentResults.tsx:188]: I considered this but aren't they synonyms? And we already use an icon of a bookmark not a pin.

Comment (JamesHuf) [reply apps/web/src/components/AssessmentStepTracker/AssessmentResults.tsx]: I've updated this with [Change some copy to use 'pin' instead of 'bookmark'](https://github.com/GCTC-NTGC/gc-digital-talent/pull/8767/commits/bd5d8656334aaf072525789e776fe3d8e36a5131) but I feels like it loses some consistency since we use "Bookmark" elsewhere.

File: apps/web/src/components/AssessmentStepTracker/AssessmentResults.tsx (hunk 3 above)
Comment (JamesHuf) [reply apps/web/src/components/AssessmentStepTracker/AssessmentResults.tsx]: I've updated this with [Change some copy to use 'pin' instead of 'bookmark'](https://github.com/GCTC-NTGC/gc-digital-talent/pull/8767/commits/bd5d8656334aaf072525789e776fe3d8e36a5131) but I feels like it loses some consistency since we use "Bookmark" elsewhere.

File: apps/web/src/components/AssessmentStepTracker/AssessmentResults.tsx (hunk 2 above)
Comment (mnigh) [reply apps/web/src/components/AssessmentStepTracker/AssessmentResults.tsx]: @GCTC-NTGC/designers thoughts?

File: apps/web/src/components/AssessmentStepTracker/AssessmentResults.tsx (hunk 3 above)
Comment (mnigh) [reply apps/web/src/components/AssessmentStepTracker/AssessmentResults.tsx]: @GCTC-NTGC/designers thoughts?

Comment (substrae) [reply apps/web/src/components/AssessmentStepTracker/AssessmentResults.tsx]: @mnigh @JamesHuf Tough to say what Jerbo's original intent here was, but from my fresh-eyes perspective, it makes more sense to use _pin_ in this context, because they're being literally pinned to the top of the list. I agree though that the icon we've chosen makes this name confusing. Let me chat with Lauren quickly to make sure I'm not modifying something very deliberate before giving you any further direction.

Comment (substrae) [reply apps/web/src/components/AssessmentStepTracker/AssessmentResults.tsx]: @mnigh @JamesHuf So I spoke with Lauren about this, and her direction was that bookmarking/pinning are both fine, but that it's important for us to keep the main requirement in mind, which is that the behaviour should be perceived as agnostic and avoid connotations of rankings/ratings/etc.

That all said, the fact that the change from _bookmark_ to pin feels very last minute (post-PR as opposed to issue hand-off) makes me lean towards keeping this as is (bookmark) until we can have a proper conversation with Jerbo about the direction he wants to take.

**Review by petertgiles — CHANGES_REQUESTED** [review]
Cool, it's looking great!  :star2:   I've noted a few things to check on.

File: apps/web/src/components/AssessmentStepTracker/utils.ts [hunk 4]
```diff
@@ -109,10 +109,46 @@ export const decisionOrder: AssessmentDecision[] = [
...
+      );
+    })
+    .map((result, index) => {
+      return {
+        ...result,
+        ordinal: index + 1,
+      };
+    });
+
+  return withOrdinal.sort((resultA, resultB) => {
```
[the remark this answers is not in the document]
Comment (JamesHuf) [reply apps/web/src/components/AssessmentStepTracker/utils.ts:150]: The map is saving the order from the first sort so I'm not sure how to do this without it. I'd love to hear if you have an improvement though.

File: api/graphql/schema.graphql [hunk 5]
```diff
@@ -1545,6 +1548,9 @@ type Mutation {
     id: ID!
     poolCandidate: UpdatePoolCandidateAsAdminInput! @spread
   ): PoolCandidate @update @guard @can(ability: "updateAsAdmin", find: "id")
+  togglePoolCandidateBookmark(id: ID!): Boolean
```
[the remark this answers is not in the document]
Comment (JamesHuf) [reply api/graphql/schema.graphql:1554]: Updated in [Change permissions for field and add comment to mutation in schema](https://github.com/GCTC-NTGC/gc-digital-talent/pull/8767/commits/06213cc4c44bfe6135fc3aa6946e8991d1eadb89)

File: api/graphql/schema.graphql [hunk 6]
```diff
@@ -267,6 +267,7 @@ type PoolCandidate {
   submittedAt: DateTime @rename(attribute: "submitted_at")
   suspendedAt: DateTime @rename(attribute: "suspended_at")
   deletedDate: DateTime @rename(attribute: "deleted_at")
+  isBookmarked: Boolean @rename(attribute: "is_bookmarked")
```
[the remark this answers is not in the document]
Comment (JamesHuf) [reply api/graphql/schema.graphql]: Updated with [Change permissions for field and add comment to mutation in schema](https://github.com/GCTC-NTGC/gc-digital-talent/pull/8767/commits/06213cc4c44bfe6135fc3aa6946e8991d1eadb89) to use `@canOnParent(ability: "viewAssessmentResults")` like we are for notes in #8812. It looks like the policy won't prevent admins from viewing these fields for their own applications - is this something we're ok with?

File: apps/web/src/components/AssessmentStepTracker/AssessmentResults.tsx [hunk 7]
```diff
@@ -88,19 +145,72 @@ const AssessmentResult = ({
...
-        <BookmarkIcon
-          {...iconStyles}
-          data-h2-color="base(gray)"
-          data-h2-flex-shrink="base(0)"
-        />
+        {isBookmarked ? (
+          <Button
+            mode="icon_only"
+            color="primary"
+            onClick={toggleBookmark}
```
[the remark this answers is not in the document]
Comment (JamesHuf) [reply apps/web/src/components/AssessmentStepTracker/AssessmentResults.tsx]: Updated in [Button updates from feedback](https://github.com/GCTC-NTGC/gc-digital-talent/pull/8767/commits/409d36684bd6668b44d7424d8ed229b1a66f598b).

File: apps/web/src/components/AssessmentStepTracker/AssessmentResults.tsx [hunk 8]
```diff
@@ -88,19 +145,72 @@ const AssessmentResult = ({
...
+                ),
+              },
+            )}
+            {...{
+              "data-h2-height": "base(x.9)",
+              "data-h2-width": "base(x.9)",
+            }}
+          />
+        ) : (
+          <Button
```
[the remark this answers is not in the document]
Comment (JamesHuf) [reply apps/web/src/components/AssessmentStepTracker/AssessmentResults.tsx]: Updated in [Button updates from feedback](https://github.com/GCTC-NTGC/gc-digital-talent/pull/8767/commits/409d36684bd6668b44d7424d8ed229b1a66f598b).

File: apps/web/src/components/AssessmentStepTracker/utils.ts [hunk 9]
```diff
@@ -109,10 +109,46 @@ export const decisionOrder: AssessmentDecision[] = [
   AssessmentDecision.Unsuccessful,
 ];
 
-export const sortResults = (
+export const sortResultsAndAddOrdinal = (
```
[the remark this answers is not in the document]
Comment (JamesHuf) [reply apps/web/src/components/AssessmentStepTracker/utils.ts:112]: Added a couple tests for this with [Add jest tests for AssessmentStepTracker](https://github.com/GCTC-NTGC/gc-digital-talent/pull/8767/commits/70e8c050f6d104c3ae06c075d4c86fff47e2a8c4).

File: api/graphql/schema.graphql (hunk 6 above)
[the remark this answers is not in the document]
Comment (petertgiles) [reply api/graphql/schema.graphql]: Yeah, I think that's fine.

File: apps/web/src/components/AssessmentStepTracker/utils.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (petertgiles) [reply apps/web/src/components/AssessmentStepTracker/utils.ts:150]: Oh, I see.  OK, I misunderstood the reason for this.  :+1:   Could you add some comments to this to help explain the reason?

File: apps/web/src/components/AssessmentStepTracker/AssessmentResults.tsx (hunk 2 above)
Comment (petertgiles) [reply apps/web/src/components/AssessmentStepTracker/AssessmentResults.tsx]: Instructions from designers: no copy changes for now

**Comment by petertgiles** [comment]
Had a quick review:
1) Comments explaining step-by-step in sort function.
2) Possible refactoring in sort function.
3) Magic number in Jest test.
4) Small unit test for just sort function in Jest testing.

**Comment by JamesHuf** [comment]
> Had a quick review:
> 
> 1. Comments explaining step-by-step in sort function.
> 2. Possible refactoring in sort function.
> 3. Magic number in Jest test.
> 4. Small unit test for just sort function in Jest testing.

I've made the changes mentioned as well as adding sorting by name since I realized I'd missed that. I'm not a huge fan of the new test though since it feels kind of hard to follow.

**Comment by JamesHuf** [comment]
Also @Jerryescandon are active members of the armed forces considered veterans in terms of priority/ordering? It looks like we treat them the same as non members right now and I'd like to make sure that's right.

File: apps/web/src/components/AssessmentStepTracker/utils.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (petertgiles) [reply apps/web/src/components/AssessmentStepTracker/utils.ts]: The refactor with the extra comments turned out really nice!

**Comment by petertgiles** [comment]
> I've made the changes mentioned as well as adding sorting by name since I realized I'd missed that. I'm not a huge fan of the new test though since it feels kind of hard to follow.

One possibility is that you can make the test more expressive by adding descriptive IDs to each element instead of just integers:
image
This makes the assertion part of the test a bit clearer:
image

**Review by petertgiles — APPROVED** [review]

**Comment by Jerryescandon** [comment]
> Also @Jerryescandon are active members of the armed forces considered veterans in terms of priority/ordering? It looks like we treat them the same as non members right now and I'd like to make sure that's right.

That sounds correct, I believe only the third option should trigger a priority change:
image

**Comment by JamesHuf** [comment]
Awesome! Thank you!
